# Conectando aos bancos de dados

##MySQL - Banco de dados COVID

In [ ]:
#Pacote necessário para ler banco de dados MySQL
!pip install PyMySQL

In [ ]:
#Importando as bibliotecas
import pymysql
import pandas as pd

In [ ]:
#Criando uma conexão com o banco de dados
conn = pymysql.connect(
    host='35.199.76.222',
    port=3306,
    user="teste-dados-leitura",
    passwd='o7c4Cc8NDeXYbAMH',
    db="teste_dados")


In [ ]:
#Criando um cursor para navegar no banco de dados
c = conn.cursor()

In [ ]:
#Verificando quantas tabelas existem no banco de dados
c.execute('SHOW TABLES')

1

In [ ]:
#Visualizando o nome da tabela para fazer a query
c.fetchall()

(('DADOS_COVID',),)

In [ ]:
#Criando o DataFrame no Python fazendo uma query SQL 
dados_covid = pd.read_sql_query("SELECT * FROM DADOS_COVID", conn)

In [ ]:
#Visualizando o banco de dados importado
dados_covid.head()

date  day  month  ...  deaths  countries_and_territories  geo_id
0  2020-03-20   20      3  ...       0                       Chad      TD
1  2020-03-21   21      3  ...       0                       Chad      TD
2  2020-03-22   22      3  ...       0                       Chad      TD
3  2020-03-23   23      3  ...       0                       Chad      TD
4  2020-03-24   24      3  ...       0                       Chad      TD

[5 rows x 10 columns]

##MongoDB - Banco de dados Multas

In [ ]:
#Instalação de pacote necessário para importar um banco de dados MongoDB
!pip install pymongo[srv]

In [ ]:
#Outro pacote necessário para importar
!pip install dnspython

In [ ]:
#Importando bibliotecas
import pymongo
import json
from pymongo import MongoClient
import dns

In [ ]:
#Criando o client para conectar com o MongoDB através da connection string disponibilizada
client = MongoClient("mongodb+srv://teste_dados_leitura:o7c4Cc8NDeXYbAMH@teste-dados.mcqmr.mongodb.net/myFirstDatabase?retryWrites=true&w=majority")

In [ ]:
#Visualizando os banco de dados do MongoDB
client.list_database_names()

['teste_dados', 'admin', 'local']

In [ ]:
#Selecionando o banco de dados que desejo trabalhar com
db = client["teste_dados"]

In [ ]:
#Verificando nome da coleção a ser utilizada
db.collection_names()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: DeprecationWarning: collection_names is deprecated. Use list_collection_names instead.
  


['multas_2020']

In [ ]:
#Criando variável com a coleção para importar para o Python
test = db.multas_2020

In [ ]:
#Importando coleção
dados_multas = pd.DataFrame(list(test.find()))

In [ ]:
#Visualizando resultado
dados_multas.head()

_id  ... quantidade_autos
0  60ab8f18ef5d383ce8403de0  ...                1
1  60ab8f18ef5d383ce8403de1  ...                2
2  60ab8f18ef5d383ce8403de2  ...                3
3  60ab8f18ef5d383ce8403de3  ...                2
4  60ab8f18ef5d383ce8403de4  ...                8

[5 rows x 7 columns]

#Tratamento dos Dados Covid

##Verificando se existem campos vazios no banco de dados

In [ ]:
#Vendo se existem valores vazios
dados_covid.isna().value_counts()

date   day    month  year   daily_confirmed_cases  daily_deaths  confirmed_cases  deaths  countries_and_territories  geo_id
False  False  False  False  False                  False         False            False   False                      False     61900
dtype: int64

##Renomeando nome das Colunas

In [ ]:
#Função para renomear as colunas
dados_covid.rename(columns={"date": "Data",
                            "day": "Dia",
                            "month": "Mês",
                            "year": "Ano",
                            "daily_confirmed_cases": "Casos Diários Confirmados",
                            "daily_deaths": "Mortes Diárias",
                            "confirmed_cases": "Casos Totais Confirmados",
                            "deaths": "Mortes Totais",
                            "countries_and_territories":"Países e Territórios",
                            "geo_id":"ID_Geográfico"},
                            inplace=True
                   )

In [ ]:
#Visualizando o resultado
dados_covid.head()

Data  Dia  Mês  ...  Mortes Totais  Países e Territórios  ID_Geográfico
0  2020-03-20   20    3  ...              0                  Chad             TD
1  2020-03-21   21    3  ...              0                  Chad             TD
2  2020-03-22   22    3  ...              0                  Chad             TD
3  2020-03-23   23    3  ...              0                  Chad             TD
4  2020-03-24   24    3  ...              0                  Chad             TD

[5 rows x 10 columns]

##Tratando a coluna Países e Territórios

In [ ]:
#Visualizando os nomes dos países
dados_covid["Países e Territórios"].unique()

array(['Chad', 'ChAd', 'ChaD', 'Cuba', 'Fiji', 'Guam', 'Iran', 'Iraq',
       'Laos', 'Mali', 'Oman', 'Peru', 'Togo', 'Aruba', 'Benin', 'Chile',
       'China', 'Congo', 'Egypt', 'Gabon', 'Ghana', 'Haiti', 'India',
       'Italy', 'Japan', 'Kenya', 'Libya', 'Malta', 'Nepal', 'Niger',
       'Qatar', 'Spain', 'Sudan', 'Syria', 'Yemen', 'Angola', 'Belize',
       'Bhutan', 'Brazil', 'BraZil', 'BraziL', 'brazil', 'bRazil',
       'BrazIl', 'Canada', 'Cyprus', 'France', 'FrAnce', 'Gambia',
       'Greece', 'Guinea', 'Guyana', 'Israel', 'Jersey', 'Jordan',
       'Kosovo', 'Kuwait', 'Latvia', 'LatVia', 'Malawi', 'Mexico',
       'Monaco', 'Norway', 'Panama', 'Poland', 'Russia', 'Rwanda',
       'Serbia', 'Sweden', 'Taiwan', 'Turkey', 'Uganda', 'Zambia',
       'Albania', 'Algeria', 'Andorra', 'Armenia', 'Austria', 'Bahamas',
       'BahAmas', 'Bahrain', 'Belarus', 'Belgium', 'Bermuda', 'Bolivia',
       'Burundi', 'Comoros', 'Croatia', 'CroatiA', 'Czechia', 'Denmark',
       'Ecuador', 'Eri

Vários países tem o nome repetido escritos de forma diferente, então vou padronizar para ficar a primeira letra maíscula e o resto minúsculo

In [ ]:
#Alterando a coluna
dados_covid["Países e Territórios"] = dados_covid["Países e Territórios"].str.title()

In [ ]:
#Visualizando o objetivo foi alcançado
dados_covid["Países e Territórios"].unique()

array(['Chad', 'Cuba', 'Fiji', 'Guam', 'Iran', 'Iraq', 'Laos', 'Mali',
       'Oman', 'Peru', 'Togo', 'Aruba', 'Benin', 'Chile', 'China',
       'Congo', 'Egypt', 'Gabon', 'Ghana', 'Haiti', 'India', 'Italy',
       'Japan', 'Kenya', 'Libya', 'Malta', 'Nepal', 'Niger', 'Qatar',
       'Spain', 'Sudan', 'Syria', 'Yemen', 'Angola', 'Belize', 'Bhutan',
       'Brazil', 'Canada', 'Cyprus', 'France', 'Gambia', 'Greece',
       'Guinea', 'Guyana', 'Israel', 'Jersey', 'Jordan', 'Kosovo',
       'Kuwait', 'Latvia', 'Malawi', 'Mexico', 'Monaco', 'Norway',
       'Panama', 'Poland', 'Russia', 'Rwanda', 'Serbia', 'Sweden',
       'Taiwan', 'Turkey', 'Uganda', 'Zambia', 'Albania', 'Algeria',
       'Andorra', 'Armenia', 'Austria', 'Bahamas', 'Bahrain', 'Belarus',
       'Belgium', 'Bermuda', 'Bolivia', 'Burundi', 'Comoros', 'Croatia',
       'Czechia', 'Denmark', 'Ecuador', 'Eritrea', 'Estonia', 'Finland',
       'Georgia', 'Germany', 'Grenada', 'Hungary', 'Iceland', 'Ireland',
       'Jamaica', 'L

##Tratando a coluna de Data

In [ ]:
#Visualizando a coluna de data
dados_covid["Data"]

0        2020-03-20
1        2020-03-21
2        2020-03-22
3        2020-03-23
4        2020-03-24
            ...    
61895    2020-02-24
61896    2020-02-26
61897    2020-03-10
61898    2020-02-20
61899    2020-02-29
Name: Data, Length: 61900, dtype: object

A data está com um formato diferente do que estamos acostumados a trabalhar, então converterei para ficar Dia/Mês/Ano

In [ ]:
import datetime as dt

In [ ]:
#Alterando a data para o padrão proposto
dados_covid['Data'] = pd.to_datetime(dados_covid['Data'], format='%Y-%m-%d').dt.strftime('%d/%m/%Y')

In [ ]:
#Visuzalizando resultado
dados_covid['Data'].head()

0    20/03/2020
1    21/03/2020
2    22/03/2020
3    23/03/2020
4    24/03/2020
Name: Data, dtype: object

##Tratando a coluna ID_Geográfico

In [ ]:
#Visualizando valores únicos da coluna ID_Geográfico
dados_covid["ID_Geográfico"].unique()

array(['TD', 'Td', 'tD', 'CU', 'FJ', 'GU', 'IR', 'IQ', 'LA', 'ML', 'OM',
       'PE', 'TG', 'AW', 'BJ', 'CL', 'CN', 'CG', 'EG', 'GA', 'GH', 'HT',
       'IN', 'IT', 'JP', 'KE', 'LY', 'MT', 'NP', 'NE', 'QA', 'ES', 'SD',
       'SY', 'YE', 'AO', 'BZ', 'BT', 'BR', 'Br', 'bR', 'CA', 'CY', 'FR',
       'GM', 'EL', 'GN', 'GY', 'IL', 'JE', 'JO', 'XK', 'KW', 'LV', 'MW',
       'MX', 'MC', 'NO', 'PA', 'PL', 'RU', 'RW', 'RS', 'SE', 'TW', 'TR',
       'UG', 'ZM', 'AL', 'DZ', 'AD', 'AM', 'AT', 'BS', 'BH', 'BY', 'BE',
       'BM', 'BO', 'BI', 'KM', 'HR', 'CZ', 'DK', 'EC', 'ER', 'EE', 'FI',
       'GE', 'DE', 'GD', 'HU', 'IS', 'IE', 'JM', 'LB', 'LS', 'LR', 'MD',
       'MA', 'MM', 'NA', 'NG', 'RO', 'SN', 'SO', 'TN', 'UA', 'UY', 'VU',
       'VN', 'AI', 'BB', 'BW', 'BG', 'KH', 'CM', 'CO', 'CW', 'DJ', 'DM',
       'SZ', 'ET', 'GG', 'VA', 'HN', 'MY', 'MV', 'MN', 'PK', 'PY', 'PT',
       'SK', 'SI', 'SR', 'TH', 'ZW', 'AR', 'AU', 'GI', 'GL', 'GT', 'ID',
       'LT', 'MU', 'NI', 'PS', 'SG', 'LK', 'VE', 'A

Existe mais de um id para o mesmo país, então formatarei para deixar todos iguais

In [ ]:
#Alterando a coluna
dados_covid["ID_Geográfico"] = dados_covid["ID_Geográfico"].str.upper()

In [ ]:
#Visualizando resultado
dados_covid["ID_Geográfico"].unique()

array(['TD', 'CU', 'FJ', 'GU', 'IR', 'IQ', 'LA', 'ML', 'OM', 'PE', 'TG',
       'AW', 'BJ', 'CL', 'CN', 'CG', 'EG', 'GA', 'GH', 'HT', 'IN', 'IT',
       'JP', 'KE', 'LY', 'MT', 'NP', 'NE', 'QA', 'ES', 'SD', 'SY', 'YE',
       'AO', 'BZ', 'BT', 'BR', 'CA', 'CY', 'FR', 'GM', 'EL', 'GN', 'GY',
       'IL', 'JE', 'JO', 'XK', 'KW', 'LV', 'MW', 'MX', 'MC', 'NO', 'PA',
       'PL', 'RU', 'RW', 'RS', 'SE', 'TW', 'TR', 'UG', 'ZM', 'AL', 'DZ',
       'AD', 'AM', 'AT', 'BS', 'BH', 'BY', 'BE', 'BM', 'BO', 'BI', 'KM',
       'HR', 'CZ', 'DK', 'EC', 'ER', 'EE', 'FI', 'GE', 'DE', 'GD', 'HU',
       'IS', 'IE', 'JM', 'LB', 'LS', 'LR', 'MD', 'MA', 'MM', 'NA', 'NG',
       'RO', 'SN', 'SO', 'TN', 'UA', 'UY', 'VU', 'VN', 'AI', 'BB', 'BW',
       'BG', 'KH', 'CM', 'CO', 'CW', 'DJ', 'DM', 'SZ', 'ET', 'GG', 'VA',
       'HN', 'MY', 'MV', 'MN', 'PK', 'PY', 'PT', 'SK', 'SI', 'SR', 'TH',
       'ZW', 'AR', 'AU', 'GI', 'GL', 'GT', 'ID', 'LT', 'MU', 'NI', 'PS',
       'SG', 'LK', 'VE', 'AZ', 'BD', 'CV', 'CR', 'K

##Observações

In [ ]:
#Verificando mais uma vez se existem tratamentos pendentes
dados_covid

Data  ...  ID_Geográfico
0      20/03/2020  ...             TD
1      21/03/2020  ...             TD
2      22/03/2020  ...             TD
3      23/03/2020  ...             TD
4      24/03/2020  ...             TD
...           ...  ...            ...
61895  24/02/2020  ...       JPG11668
61896  26/02/2020  ...       JPG11668
61897  10/03/2020  ...       JPG11668
61898  20/02/2020  ...       JPG11668
61899  29/02/2020  ...       JPG11668

[61900 rows x 10 columns]

Após olhar mais uma vez para o banco de dados para verificar se existia alguma alteração a ser feita, foi observado que na linha 61897 existe um número negativo para Casos Diários Confirmados, então fui investigar.

In [ ]:
#Pesquisando linhas que possuem casos diários confirmados negativos
dados_covid.loc[(dados_covid['Casos Diários Confirmados'] < 0).values]

Data  ...  ID_Geográfico
3544   20/05/2020  ...             BJ
6290   20/06/2020  ...             IT
8695   19/04/2020  ...             ES
8730   25/05/2020  ...             ES
11667  03/06/2020  ...             FR
13605  11/09/2020  ...             JE
13911  22/07/2020  ...             JO
18706  21/05/2020  ...             UG
18718  02/06/2020  ...             UG
24195  07/05/2020  ...             EC
24197  09/05/2020  ...             EC
24199  12/05/2020  ...             EC
24297  07/09/2020  ...             EC
37791  03/05/2020  ...             PT
41583  29/04/2020  ...             LT
45303  28/08/2020  ...             LU
46926  11/05/2020  ...             SM
61897  10/03/2020  ...       JPG11668

[18 rows x 10 columns]

Também pode-se observar que existem Mortes Diárias negativas. Não sei ao certo a origem dos dados, então não tenho um dicionário para saber o real motivo. Mas eu deduzo que sejam números que foram feitos de forma errada na data anterior, e no dia seguinte foi corrigido, já que nos Casos Totais Confirmados e nas Mortes Confirmadas, não existem números negativos.

In [ ]:
#Comprovando que não existem números negativos nos Casos Totais Confirmados
(dados_covid['Casos Totais Confirmados'] < 0).values.any()

False

In [ ]:
#Comprovando que não existem números negativos nas Mortes Totais
(dados_covid['Mortes Totais'] < 0).values.any()

False

Isto é algo que será tem que ser levado em conta na hora das análises.

 Agora o banco de Dados COVID está pronto para ser analisado.

#Tratamento dos Dados Multas

##Verificando se existem campos vazios no banco de dados

In [ ]:
#Visualizando se existem valores negativos
dados_multas.isna().value_counts()

_id    escopo_autuacao  mes    uf     amparo_legal  descricao_infracao  quantidade_autos
False  False            False  False  False         False               False               334
dtype: int64

##Renomeando nome das colunas

In [ ]:
#Função para renomear colunas
dados_multas.rename(columns={"_id": "ID",
                            "escopo_autuacao": "Escopo Autuação",
                            "mes": "Mês",
                            "uf": "UF",
                            "amparo_legal": "Amparo Legal",
                            "descricao_infracao": "Descrição Infração",
                            "quantidade_autos": "Quantidade Autos"},
                            inplace=True
                   )

In [ ]:
#Visualizando o resultado
dados_multas.head()

ID  ... Quantidade Autos
0  60ab8f18ef5d383ce8403de0  ...                1
1  60ab8f18ef5d383ce8403de1  ...                2
2  60ab8f18ef5d383ce8403de2  ...                3
3  60ab8f18ef5d383ce8403de3  ...                2
4  60ab8f18ef5d383ce8403de4  ...                8

[5 rows x 7 columns]

##Tratando a coluna Mês

In [ ]:
#Visualizando como está a coluna Mês
dados_multas["Mês"].unique()

array(['JANEIRO', 'JANeIRO', 'JaNEIRO', 'FEVEREIRO', 'MARÇO', 'ABRIL',
       'MAIO', 'JUNHO', 'JULHO', 'AGOSTO', 'SETEMBRO', 'OUTUBRO',
       'NOVEMBRO', 'DEZEMBRO'], dtype=object)

Coluna está com nomes repetidos de formas diferentes. Eu iria passar somente para um padrão com a primeira letra maiúscula e o restante minúsculo (Ex: JaNeiRO -> Janeiro). No entanto quero poder relacionar os bancos de dados do Covid com o de Multas. Então transformarei para números (Ex: Janeiro -> 1).

In [ ]:
#Passando todos os meses para o padrão de primeira letra maiúscula e o resto minúsculo
dados_multas["Mês"] = dados_multas["Mês"].str.title()

In [ ]:
#Visualizando coluna
dados_multas["Mês"]

0       Janeiro
1       Janeiro
2       Janeiro
3       Janeiro
4       Janeiro
         ...   
329    Dezembro
330    Dezembro
331    Dezembro
332    Dezembro
333    Dezembro
Name: Mês, Length: 334, dtype: object

In [ ]:
#Criando função para transformar o nome do mês em número inteiro
def mes_nome_para_numero(mes, numero):
  dados_multas["Mês"].replace(mes,numero,inplace=True)

In [ ]:
##Testando função
mes_nome_para_numero('Janeiro',1)

In [ ]:
##Verificando
dados_multas["Mês"]

0             1
1             1
2             1
3             1
4             1
         ...   
329    Dezembro
330    Dezembro
331    Dezembro
332    Dezembro
333    Dezembro
Name: Mês, Length: 334, dtype: object

In [ ]:
#Aplicando a função mes_nome_para_numero para o restante dos meses
mes_nome_para_numero('Fevereiro',2)
mes_nome_para_numero('Março',3)
mes_nome_para_numero('Abril',4)
mes_nome_para_numero('Maio',5)
mes_nome_para_numero('Junho',6)
mes_nome_para_numero('Julho',7)
mes_nome_para_numero('Agosto',8)
mes_nome_para_numero('Setembro',9)
mes_nome_para_numero('Outubro',10)
mes_nome_para_numero('Novembro',11)
mes_nome_para_numero('Dezembro',12)

In [ ]:
#Verificando se todos os meses foram alterados com sucesso
dados_multas["Mês"].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

## Tratando a coluna UF

In [ ]:
#Observando coluna UF
dados_multas["UF"].unique()

array(['Es', 'Mg', 'MG', 'MS', 'RJ', 'SC', 'ES', 'BA', 'MT', 'PR', 'RS',
       'SP', 'GO'], dtype=object)

Temos que tratar os estados que não estão com as duas letras em maiúsculo para ficar no padrão.

In [ ]:
#Transformando todas as letras para maiúscula
dados_multas["UF"] = dados_multas["UF"].str.upper()

In [ ]:
#Agora todos estão no padrão
dados_multas["UF"].unique()

array(['ES', 'MG', 'MS', 'RJ', 'SC', 'BA', 'MT', 'PR', 'RS', 'SP', 'GO'],
      dtype=object)

##Tratando a coluna Quantidade Autos

Ao tentar observar se a coluna Quantidade Autos possui números negativos utilizando a linha de código: /n
(dados_multas["Quantidade Autos"] < 0).values.any() /n
Ocorreu o erro '<' not supported between instances of 'str' and 'int, ou seja, os números estão como o tipo String e não Int, então precisamos converter.

In [ ]:
#Convertendo as Strings para Int
dados_multas["Quantidade Autos"] = dados_multas["Quantidade Autos"].astype(int)

In [ ]:
#Verificando se existem valores negativos na coluna Quantidade Autos
(dados_multas["Quantidade Autos"] < 0).values.any()

False

##Tratando a coluna Amparo Legal

In [ ]:
#Ao dar uma olhada nos dados novamente, observei que existem linhas que estão preenchidas com colchetes []
dados_multas['Amparo Legal'].head(30)

0     Lei 9503/97 Lei 10561/02 Res. CONTRAN nº 210/0...
1     Lei 9503/97 Lei 10561/02 Res. CONTRAN nº 210/0...
2     Lei 9503/97 Lei 10561/02 Res. CONTRAN nº 210/0...
3     Lei 9503/97 Lei 10561/02 Res. CONTRAN nº 210/0...
4     Lei 9503/97 Lei 10561/02 Res. CONTRAN nº 210/0...
5     Lei 9503/97 Lei 10561/02 Res. CONTRAN nº 210/0...
6     Lei 9503/97 Lei 10561/02 Res. CONTRAN nº 210/0...
7     Lei 9503/97 Lei 10561/02 Res. CONTRAN nº 210/0...
8     Lei 9503 DE 23/09/1997 Transpor, sem autorizaç...
9     Lei 9503 DE 23/09/1997 Transpor, sem autorizaç...
10    Lei 9503/97 Lei 10561/02 Res. CONTRAN nº 210/0...
11    Lei 9503/97 Lei 10561/02 Res. CONTRAN nº 210/0...
12    Lei 9503/97 Lei 10561/02 Res. CONTRAN nº 210/0...
13    Lei 9503/97 Lei 10561/02 Res. CONTRAN nº 210/0...
14    Lei 9503/97 Lei 10561/02 Res. CONTRAN nº 210/0...
15    Lei 9503/97 Lei 10561/02 Res. CONTRAN nº 210/0...
16    Lei 9503/97 Lei 10561/02 Res. CONTRAN nº 210/0...
17    Lei 9503/97 Lei 10561/02 Res. CONTRAN nº 2

In [ ]:
#Buscando as linhas que estão preenchidas com os colchetes[]
dados_multas[dados_multas['Amparo Legal'].str.len() < 4]

ID  ... Quantidade Autos
26   60ab8f18ef5d383ce8403dfa  ...                1
187  60ab8f18ef5d383ce8403e9b  ...                1

[2 rows x 7 columns]

In [ ]:
#Ao pesquisar sobre a infração, vi que ela era amparada pela Resolução ANTT Nº 4282 DE 17/03/2014, então irei substituir os colchetes
dados_multas.at[26, 'Amparo Legal'] = "Resolução ANTT Nº 4282 DE 17/03/2014"

In [ ]:
#Não estava conseguindo utilizar o método replace para substituir os colchetes, então utilizei o método at
dados_multas.at[187, 'Amparo Legal'] = "Resolução ANTT Nº 4282 DE 17/03/2014"

In [ ]:
#Visualizando resultados
dados_multas['Amparo Legal'].loc[26]

'Resolução ANTT Nº 4282 DE 17/03/2014'

In [ ]:
#Visualizando resultados
dados_multas['Amparo Legal'].loc[187]

'Resolução ANTT Nº 4282 DE 17/03/2014'

Agora os dados estão prontos para serem utilizados para análise, como não sei se a empresa permite importar estes dados para o MySQL e para o MongoDB disponibilizado, irei exportar para CSV e continuar o trabalho a partir destes arquivos.

In [ ]:
#Criando arquivo CSV para banco de dados Multas
dados_multas.to_csv(r'dados_multas.csv', index=False, encoding='utf-8')

In [ ]:
#Criando arquivo CSV para banco de dados Covid
dados_covid.to_csv(r'dados_covid.csv', index=False, encoding='utf-8')

#Conclusão

Desta forma, finalizo o processo de tratamento dos dados, podemos observar que existiam vários problemas, mesmo que pequenos, que precisavam ser resolvidos para fazermos uma análise precisa dos dados, colocando todas as variáveis nos padrões e convertendo para o tipo certo.